Source/Steam/ServerManager.cpp

```diff
@@ -0,0 +1,96 @@
+// SPDX-License-Identifier: MIT
+#include "PortabilityInfo.h"
+#include "Common/FEXServerClient.h"
+
+#include <cstdio>
+#include <unistd.h>
+#include <poll.h>
+
+void MsgHandler(LogMan::DebugLevels Level, const char* Message) {
+  const auto Style = fmt::text_style {};
+  const auto Output = fextl::fmt::format("{} {}\n", fmt::styled(LogMan::DebugLevelStr(Level), Style), Message);
+  write(STDERR_FILENO, Output.c_str(), Output.size());
+  fsync(STDERR_FILENO);
+}
+
+void AssertHandler(const char* Message) {
+  return MsgHandler(LogMan::ASSERT, Message);
+}
+
+void SignalPVToContinue() {
+  // Tell pressure-vessel that the startup was a success.
+  const auto ReadyMsg = "READY=1\n";
+  write(STDOUT_FILENO, ReadyMsg, strlen(ReadyMsg));
+
+  // pressure-vessel is waiting for EOF on STDOUT from this process to ensure it can run FEX processes.
+  // dup2 atomically replaces stdout with a copy of stderr to achieve this.
+  dup2(STDERR_FILENO, STDOUT_FILENO);
+}
+
+struct PipesType {
+  int read_pipe {-1};
+  int write_pipe {-1};
+};
+
+PipesType get_pipe() {
+  PipesType pipes {};
+  pipe(&pipes.read_pipe);
+  return pipes;
+}
+
+int main(int argc, const char** argv, char** const envp) {
+  LogMan::Throw::InstallHandler(AssertHandler);
+  LogMan::Msg::InstallHandler(MsgHandler);
+
+  const auto PortableInfo = FEX::ReadPortabilityInformation();
+  FEX::Config::LoadConfig({}, envp, PortableInfo);
+
+  // Reload the meta layer
+  FEXCore::Config::ReloadMetaLayer();
+
+  auto pipes = get_pipe();
+
+  // Set the write side to close on exec.
+  fcntl(pipes.write_pipe, F_SETFD, FD_CLOEXEC);
+
+  // Give the read end of the pipe to FEXServer.
+  auto ServerFD = FEXServerClient::StartServer(PortableInfo.InterpreterPath, pipes.read_pipe);
```

Comment:
What does the server do with the read end of the pipe? Does it, similarly, wait until EOF is reached and then exit?

If yes, then instead of implementing your own pipe, you could probably just `dup()` the server manager's original `STDIN_FILENO`, pass the duplicate down to the server, and close both stdin and the duplicate in the server manager.

Reply:
Oh, but then you wouldn't know when the server manager can exit, so you'd probably have to `execve()` the server (replacing yourself with the server) rather than having run-and-wait logic. That would save us a process, but it's a rather different execution model, so I can understand if you'd rather not.